File: nasa_frontend/lib/models/fireball.dart
class Fireball {
  final int id;
  final DateTime eventDate;
  final double? latitude;
  final double? longitude;
  final double? altitudeKm;
  final double? totalImpactEnergyKt;
  final double? velocityKmS;

  Fireball({
    required this.id,
    required this.eventDate,
    this.latitude,
    this.longitude,
    this.altitudeKm,
    this.totalImpactEnergyKt,
    this.velocityKmS,
  });

  factory Fireball.fromJson(Map<String, dynamic> json) {
    return Fireball(
      id: json['id'] as int,
      eventDate: DateTime.parse(json['eventDate'] as String),
      latitude: json['latitude'] != null ? (json['latitude'] as num).toDouble() : null,
      longitude: json['longitude'] != null ? (json['longitude'] as num).toDouble() : null,
      altitudeKm: json['altitudeKm'] != null ? (json['altitudeKm'] as num).toDouble() : null,
      totalImpactEnergyKt: json['totalImpactEnergyKt'] != null
          ? (json['totalImpactEnergyKt'] as num).toDouble()
          : null,
      velocityKmS: json['velocityKmS'] != null
          ? (json['velocityKmS'] as num).toDouble()
          : null,
    );
  }

  bool get hasLocation => latitude != null && longitude != null;
}

File: nasa_backend/src/main/java/it/polimi/nasa/nasabackend/entity/Fireball.java
package it.polimi.nasa.nasabackend.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "fireball", indexes = {
        @Index(name = "idx_fireball_date", columnList = "eventDate"),
        @Index(name = "idx_fireball_location", columnList = "latitude,longitude")
})
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Fireball {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private LocalDateTime eventDate;

    // Geographic data
    @Column(precision = 9, scale = 6)
    private BigDecimal latitude;

    @Column(precision = 9, scale = 6)
    private BigDecimal longitude;

    @Column(precision = 8, scale = 2)
    private BigDecimal altitudeKm;

    // Physical properties
    @Column(precision = 15, scale = 2)
    private BigDecimal totalRadiatedEnergyJ;

    @Column(precision = 10, scale = 4)
    private BigDecimal totalImpactEnergyKt;

    @Column(precision = 8, scale = 2)
    private BigDecimal velocityKmS;

    // Velocity components
    @Column(precision = 10, scale = 2)
    private BigDecimal velocityVx;

    @Column(precision = 10, scale = 2)
    private BigDecimal velocityVy;

    @Column(precision = 10, scale = 2)
    private BigDecimal velocityVz;

    // Metadata
    @ManyToOne
    @JoinColumn(name = "api_source_id")
    private ApiSource apiSource;

    @Column(columnDefinition = "TEXT")
    private String rawData;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}